arch/risc-v/src/mpfs/mpfs_can.c

```diff
@@ -0,0 +1,2953 @@
+ *
+ * Input Parameters:
+ *  priv  - Pointer to the private CAN driver state structure
+ *
+ * Returned Value:
+ *  None
+ *
+ * Assumptions:
+ *  Global interrupts are disabled by the watchdog logic.
+ *  We are not in an interrupt context so that we can lock the network.
+ *
+ ****************************************************************************/
+
+static void mpfs_txdone_work(void *arg)
+{
+  mpfs_can_instance_t *priv = (mpfs_can_instance_t *)arg;
+
+#ifdef CONFIG_DEBUG_CAN_INFO
+  caninfo("TX done interrupt received\n");
+#endif
+
+  /* There should be space for a new TX in any event.  Poll the network for
+   * new XMIT data
+   */
+
+  net_lock();
+  devif_poll(&priv->dev, mpfs_txpoll);
+  net_unlock();
+}
+
+/****************************************************************************
+ * Name: mpfs_transmit
+ *
+ * Description:
+ *  Start hardware transmission.  Called either from the txdone interrupt
+ *  handling or from watchdog based polling.
+ *
+ * Input Parameters:
+ *  priv  - Pointer to the private CAN driver state structure
+ *
+ * Returned Value:
+ *  Zero (CAN_OK) on success; a negated errno on failure
+ *
+ * Assumptions:
+ *  May or may not be called from an interrupt handler.  In either case,
+ *  global interrupts are disabled, either explicitly or indirectly through
+ *  interrupt handling logic.
+ *
+ ****************************************************************************/
+
+static int mpfs_transmit(mpfs_can_instance_t *priv)
+{
+  uint8_t ret;
+  uint8_t i;
+
+  /* Retrieve the CAN 2.0B frame from network device buffer */
+
+  struct can_frame *cf = (struct can_frame *)priv->dev.d_buf;
+
+  /* Fill the CAN msg object to be sent */
+
+  pmpfs_can_msgobject_t pmsg = priv->tx_msg;
+
+  /* CAN TX msg control command */
+
+  pmsg->msg_ctrl = ((cf->can_id & CAN_EFF_FLAG) ?
+                   MPFS_CAN_TX_MSG_CTRL_CMD_IDE : 0)
+                   | ((cf->can_dlc << MPFS_CAN_TX_MSG_CTRL_CMD_DLC_SHIFT)
+                   & MPFS_CAN_TX_MSG_CTRL_CMD_DLC)
+                   | ((cf->can_id & CAN_RTR_FLAG) ?
+                   MPFS_CAN_TX_MSG_CTRL_CMD_RTR : 0);
+
+  /* CAN ID */
+
+  pmsg->id = mpfs_can_canid_to_msgid(cf->can_id);
+
+  /* CAN data */
+
+  if (!(pmsg->msg_ctrl & MPFS_CAN_TX_MSG_CTRL_CMD_RTR))
+    {
+      for (i = 0; i < cf->can_dlc; i++)
+        {
+          if (i < 4)
+            {
+              pmsg->data_high |= (cf->data[i] << ((3 - i) * 8));
+            }
+          else
+            {
+              pmsg->data_low |= (cf->data[i] << ((7 - i) * 8));
+            }
+        }
+    }
+
+  /* Insert CAN msg object into available TX bf */
+
+  if (CAN_VALID_MSG != (ret = mpfs_can_send_message(priv)))
+    {
+      canwarn("Failed to send CAN frame due to %s\n",
+              ret == CAN_INVALID_BUFFER ? "invalid buffer" :
+              ret == CAN_NO_MSG ? "no TX buffer available" : "unknown err");
+      return CAN_OK;
```

Comment:
Why return CAN_OK in case of failure?

Reply:
yes, err return handling of mpfs_transmit will be fixed. thanks!